### PR TITLE
Feature/160

### PR DESCRIPTION
**Pull request by ArmaanKatyal**
In this pull request:

- Added support for the JWT key for user session authentication.

### CODE REVIEW

File: backend/src/routes/Vendors.js [hunk 1]
```diff
@@ -53,7 +53,7 @@ const router = express.Router()
  *               items:
  *                 $ref: '#/components/schemas/Vendor'
  */
-router.get('/', verifyApiKey, vendorController.getAll)
+router.get('/', verifyApiKey, validateJwtToken, vendorController.getAll)
```
Comment (dimassudjito) [inline backend/src/routes/Vendors.js:56]: we don't need Jwt token to read vendors (since you don't need to login)

**Review by dimassudjito — CHANGES_REQUESTED** [review]
Minor change needed: remove jwt middleware from vendors endpoint

**Comment by dimassudjito** [comment]
@ArmaanKatyal @TienDucHo 
Sorry, I forgot about the things that we discuss two weeks ago so correct me if I'm wrong:
1. The mobile app would log in users through google OAuth and get their email, name and picture. When they make query to /login to our backend, they'll be sending this data. They don't need to send idToken or any token from google.
2. Our backend will verify/create the user in the database and return data containing information about the user but also the JWT token which expires in 30 days.
3. The mobile app will use JTW token to make request to endpoint such as student (e.g., to update my icard page)

Question for @TienDucHo: do we get email, name, and picture directly when someone login or do we need to fetch it again from another google API? and if so, is there a performance issue with this method?

**Comment by ArmaanKatyal** [comment]
> @ArmaanKatyal @TienDucHo Sorry, I forgot about the things that we discuss two weeks ago so correct me if I'm wrong:
> 
> 1. The mobile app would log in users through google OAuth and get their email, name and picture. When they make query to /login to our backend, they'll be sending this data. They don't need to send idToken or any token from google.
> 2. Our backend will verify/create the user in the database and return data containing information about the user but also the JWT token which expires in 30 days.
> 3. The mobile app will use JTW token to make request to endpoint such as student (e.g., to update my icard page)
> 
> Question for @TienDucHo: do we get email, name, and picture directly when someone login or do we need to fetch it again from another google API? and if so, is there a performance issue with this method?

we don't need to fetch all the details from google again as they are stored in our database when the user login for the first time. It will not create any performance issues. Yep, when the user logs in again he doesn't need the Id token as we have the jwt key there to verify.

**Comment by ArmaanKatyal** [comment]
I will also add a route where duc can make requests to check if the jwt key is still valid or not. if not then the user has to log in again using the google id token. if jwt key is still valid then the user can access the data. Then we can merge this pull request.

Give me few more days. I will add this feature and test for the response time aswell.

**Comment by dimassudjito** [comment]
> > @ArmaanKatyal @TienDucHo Sorry, I forgot about the things that we discuss two weeks ago so correct me if I'm wrong:
> > 
> > 1. The mobile app would log in users through google OAuth and get their email, name and picture. When they make query to /login to our backend, they'll be sending this data. They don't need to send idToken or any token from google.
> > 2. Our backend will verify/create the user in the database and return data containing information about the user but also the JWT token which expires in 30 days.
> > 3. The mobile app will use JTW token to make request to endpoint such as student (e.g., to update my icard page)
> > 
> > Question for @TienDucHo: do we get email, name, and picture directly when someone login or do we need to fetch it again from another google API? and if so, is there a performance issue with this method?
> 
> we don't need to fetch all the details from google again as they are stored in our database when the user login for the first time. It will not create any performance issues. Yep, when the user logs in again he doesn't need the Id token as we have the jwt key there to verify.

@ArmaanKatyal @TienDucHo 
Yes, but when they first login we would need to get the name and picture from google and I was wondering how this works (I now the first time duc made this, it takes like a minute to process and then suddenly it just got fast hahah)

**Comment by dimassudjito** [comment]
> I will also add a route where duc can make requests to check if the jwt key is still valid or not. if not then the user has to log in again using the google id token. if jwt key is still valid then the user can access the data. Then we can merge this pull request.
> 
> Give me few more days. I will add this feature and test for the response time aswell.

@ArmaanKatyal 
Wouldn't it be better if the middleware that check jwt token also checks if the token has expired. If it has expired, it would send error message letting the app knows that the error is because token has expired

File: backend/src/services/verifyToken.js [hunk 2]
```diff
@@ -64,8 +64,24 @@ function checkAuthenticated(req, res, next) {
...
+const validateJwtToken = async(req, res, next) => {
+    const token = req.header('jwt-token')
+    if (!token)
+        return res.status(401).json({ message: 'Access denied. No token provided.'})
+    
+    try {
+        const verified = await jwt.verify(token, process.env.JWT_SECRET)
+        req.user = verified
+        next()
+    } catch (err) {
```
Comment (dimassudjito) [inline backend/src/services/verifyToken.js:76]: does jwt.verify() automatically check for expiry date of the token? we just need to make sure to throw descriptive error so we know when the token is expired and when it's just invalid

**Review by dimassudjito — CHANGES_REQUESTED** [review]

**Comment by dimassudjito** [comment]
@ArmaanKatyal @TienDucHo 

Ok, I just talked with Duc on the standup meeting and I'm clear on how things work now. So, when the app first login, they will pass ID token that will be verified by the backend to get user information. So, the app doesn't have to pass user data such as name and URL picture. In another query like upload image and my icard data, the app will pass JWT which will authenticate their access to the backend.

Let me know if I'm accurate and I can create a documentation for this for the new dev.

**Comment by dimassudjito** [comment]
Sorry for my confusion hahah, I missed the part where the login/ route still used the checkAuthenticated middleware

File: backend/src/routes/Vendors.js (hunk 1 above)
Comment (dimassudjito) [reply backend/src/routes/Vendors.js:56]: @ArmaanKatyal can we fix this before we merge this PR

**Review by dimassudjito — CHANGES_REQUESTED** [review]

**Review by dimassudjito — APPROVED** [review]
Looks good! Please merge when you approve @MoyinF @yismailuofa

**Pull request merged**